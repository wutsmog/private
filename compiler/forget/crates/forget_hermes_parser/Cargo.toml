[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }