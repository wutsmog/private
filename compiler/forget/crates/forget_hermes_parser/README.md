# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.