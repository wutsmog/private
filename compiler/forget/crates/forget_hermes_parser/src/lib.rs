mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
