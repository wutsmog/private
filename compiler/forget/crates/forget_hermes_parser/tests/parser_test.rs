use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
