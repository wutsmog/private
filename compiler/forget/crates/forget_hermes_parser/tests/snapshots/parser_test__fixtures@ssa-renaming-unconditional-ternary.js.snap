---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/ssa-renaming-unconditional-ternary.js
---
Input:
function foo(props) {
  let x = [];
  x.push(props.bar);
  props.cond
    ? ((x = {}), (x = []), x.push(props.foo))
    : ((x = []), (x = []), x.push(props.bar));
  return x;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "foo",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 9,
          "end": 12
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 13,
            "end": 18
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "let",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 28,
                    "end": 29
                  }
                },
                "init": {
                  "type": "ArrayExpression",
                  "elements": [],
                  "loc": null,
                  "range": {
                    "start": 32,
                    "end": 34
                  }
                },
                "loc": null,
                "range": {
                  "start": 28,
                  "end": 34
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 24,
              "end": 35
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 38,
                    "end": 39
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "push",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 40,
                    "end": 44
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 38,
                  "end": 44
                }
              },
              "arguments": [
                {
                  "type": "MemberExpression",
                  "object": {
                    "type": "Identifier",
                    "name": "props",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 45,
                      "end": 50
                    }
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "bar",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 51,
                      "end": 54
                    }
                  },
                  "computed": false,
                  "loc": null,
                  "range": {
                    "start": 45,
                    "end": 54
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 38,
                "end": 55
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 38,
              "end": 56
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "ConditionalExpression",
              "test": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "props",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 59,
                    "end": 64
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "cond",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 65,
                    "end": 69
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 59,
                  "end": 69
                }
              },
              "alternate": {
                "type": "SequenceExpression",
                "expressions": [
                  {
                    "type": "AssignmentExpression",
                    "operator": "=",
                    "left": {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 124,
                        "end": 125
                      }
                    },
                    "right": {
                      "type": "ArrayExpression",
                      "elements": [],
                      "loc": null,
                      "range": {
                        "start": 128,
                        "end": 130
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 124,
                      "end": 130
                    }
                  },
                  {
                    "type": "AssignmentExpression",
                    "operator": "=",
                    "left": {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 134,
                        "end": 135
                      }
                    },
                    "right": {
                      "type": "ArrayExpression",
                      "elements": [],
                      "loc": null,
                      "range": {
                        "start": 138,
                        "end": 140
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 134,
                      "end": 140
                    }
                  },
                  {
                    "type": "CallExpression",
                    "callee": {
                      "type": "MemberExpression",
                      "object": {
                        "type": "Identifier",
                        "name": "x",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 143,
                          "end": 144
                        }
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "push",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 145,
                          "end": 149
                        }
                      },
                      "computed": false,
                      "loc": null,
                      "range": {
                        "start": 143,
                        "end": 149
                      }
                    },
                    "arguments": [
                      {
                        "type": "MemberExpression",
                        "object": {
                          "type": "Identifier",
                          "name": "props",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 150,
                            "end": 155
                          }
                        },
                        "property": {
                          "type": "Identifier",
                          "name": "bar",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 156,
                            "end": 159
                          }
                        },
                        "computed": false,
                        "loc": null,
                        "range": {
                          "start": 150,
                          "end": 159
                        }
                      }
                    ],
                    "loc": null,
                    "range": {
                      "start": 143,
                      "end": 160
                    }
                  }
                ],
                "loc": null,
                "range": {
                  "start": 123,
                  "end": 160
                }
              },
              "consequent": {
                "type": "SequenceExpression",
                "expressions": [
                  {
                    "type": "AssignmentExpression",
                    "operator": "=",
                    "left": {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 78,
                        "end": 79
                      }
                    },
                    "right": {
                      "type": "ObjectExpression",
                      "properties": [],
                      "loc": null,
                      "range": {
                        "start": 82,
                        "end": 84
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 78,
                      "end": 84
                    }
                  },
                  {
                    "type": "AssignmentExpression",
                    "operator": "=",
                    "left": {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 88,
                        "end": 89
                      }
                    },
                    "right": {
                      "type": "ArrayExpression",
                      "elements": [],
                      "loc": null,
                      "range": {
                        "start": 92,
                        "end": 94
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 88,
                      "end": 94
                    }
                  },
                  {
                    "type": "CallExpression",
                    "callee": {
                      "type": "MemberExpression",
                      "object": {
                        "type": "Identifier",
                        "name": "x",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 97,
                          "end": 98
                        }
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "push",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 99,
                          "end": 103
                        }
                      },
                      "computed": false,
                      "loc": null,
                      "range": {
                        "start": 97,
                        "end": 103
                      }
                    },
                    "arguments": [
                      {
                        "type": "MemberExpression",
                        "object": {
                          "type": "Identifier",
                          "name": "props",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 104,
                            "end": 109
                          }
                        },
                        "property": {
                          "type": "Identifier",
                          "name": "foo",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 110,
                            "end": 113
                          }
                        },
                        "computed": false,
                        "loc": null,
                        "range": {
                          "start": 104,
                          "end": 113
                        }
                      }
                    ],
                    "loc": null,
                    "range": {
                      "start": 97,
                      "end": 114
                    }
                  }
                ],
                "loc": null,
                "range": {
                  "start": 77,
                  "end": 114
                }
              },
              "loc": null,
              "range": {
                "start": 59,
                "end": 161
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 59,
              "end": 162
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "Identifier",
              "name": "x",
              "typeAnnotation": null,
              "loc": null,
              "range": {
                "start": 172,
                "end": 173
              }
            },
            "loc": null,
            "range": {
              "start": 165,
              "end": 174
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 20,
          "end": 176
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 0,
        "end": 176
      },
      "loc": null,
      "range": {
        "start": 0,
        "end": 176
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 0,
    "end": 176
  }
}
